### PR TITLE
First refactor of compaction

### PR DESCRIPTION
**Pull request by techdocsmith**
#10897

### First pass refactor / update of compaction docs

Updates to "Data management" topic as follows:
- Adds an introduction that describes the content in the topic.
- Removes a duplicated section about "Schema changes" and leaves it in design/segments.md

Adds a new topic "Compaction" that defines compaction and automatic compaction as a strategy for segment optimization.

- Use cases for compaction. Data handling with compaction. Describes compaction task configuration."
- This includes the behavior for #10856
- This includes the behavior for #10900

Repairs links for the refactor above.

This PR doesn't handle the remaining task of identifying reindexing and compaction as data management tasks for existing data and comparing the use cases between the two. This should come in a subsequent PR.

cc: @maytasm, @suneet-s , @loquisgon , @sthetland 

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.

### CODE REVIEW

File: docs/configuration/index.md [hunk 1]
```diff
@@ -820,24 +820,24 @@ A description of the compaction config is:
...
 }
 ```
 
-Note that compaction tasks can fail if their locks are revoked by other tasks of higher priorities.
-Since realtime tasks have a higher priority than compaction task by default,
-it can be problematic if there are frequent conflicts between compaction tasks and realtime tasks.
-If this is the case, the coordinator's automatic compaction might get stuck because of frequent compaction task failures.
-This kind of problem may happen especially in Kafka/Kinesis indexing systems which allow late data arrival.
-If you see this problem, it's recommended to set `skipOffsetFromLatest` to some large enough value to avoid such conflicts between compaction tasks and realtime tasks.
+Compaction tasks fail when higher priority tasks cause Druid to revokes their locks. By default, realtime tasks like ingestion have a higher priority than compaction tasks. Therefore frequent conflicts between compaction tasks and realtime tasks can cause the coordinator's automatic compaction to get stuck.
```
Comment (2bethere) [inline docs/configuration/index.md:839]: Nice! This is way clearer.

File: docs/ingestion/compaction.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+## Types of segment compaction
+You can configure the Druid Coordinator to perform automatic compaction, also called auto-compaction, for a datasource. Using a segment search policy, the coordinator periodically identifies segments for compaction starting with the newest to oldest. When segments can benefit from compaction, the coordinator automatically submits a compaction task. 
+
+Automatic compaction works in most use cases and should be your first option. To learn more about automatic compaction, see [Compacting Segments](../design/coordinator.md#compacting-segments).
+
+In cases where you require more control over compaction, you can manually submit compaction tasks. For example:
+- Automatic compaction is too slow.
+- You want to force compaction for a specific time range.
+- Compacting recent data before older data suboptimal is suboptimal in your environment.
```
Comment (2bethere) [inline docs/ingestion/compaction.md:38]:
```suggestion
- You want to compact data in non-chronological order.
```

File: docs/ingestion/compaction.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+### Segment granularity handling
+
+Unless you modify the segment granularity in the [granularity spec](#compaction-granularity-spec), Druid attempts to retain the granularity for the compacted segments. When segments have different segment granularities with no overlap in interval Druid creates a separate compaction task for each to retain the segment granularity in the compacted segment. If segments have different segment granularities before compaction but there is some overlap in interval, Druid attempts find start and end of the overlapping interval and uses the closest segment granularity level for the compacted segment.
```
Comment (2bethere) [inline docs/ingestion/compaction.md:50]: This is unclear to me. Specifically, 
1. "segments have different segment granularities before compaction but there is some overlap in interval", I think an example can help.
2. "Druid attempts find start and end of the overlapping interval", I think more detail about why and how Druid do this attempt can help.
3. " uses the closest segment granularity level". Maybe an example here as well?

File: docs/ingestion/compaction.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+### Query granularity handling
+
+Unless you modify the query granularity in the [granularity spec](#compaction-granularity-spec), Druid retains the query granularity for the compacted segments. If segments have different query granularities before compaction, Druid chooses the finest level of granularity for the resulting compacted segment. For example if a compaction task combines two segments, one with day query granularity and one with minute query granularity, the resulting segment uses minute query granularity.
+
+> In Apache Druid 0.21.0 and prior, Druid sets the granularity for compacted segments to the default granularity of `NONE` regardless of the query granularity of the original segments.
+
+If you configure query granularity in compaction to go from a finer granularity like month to a coarser query granularity like year, then Druid overshadows the original segment with finer granularity. Because the new segments have a coarser granularity, running a kill task to remove the overshadowed segments for those intervals will cause you to permanently lose the finer granularity data.
```
Comment (2bethere) [inline docs/ingestion/compaction.md:58]:
```suggestion
If you configure query granularity in compaction to go from a finer granularity like month to a coarser query granularity like year, then Druid overshadows the original segment with coarser granularity. Because the new segments have a coarser granularity, running a kill task to remove the overshadowed segments for those intervals will cause you to permanently lose the finer granularity data.
```

Not sure if this is what you meant

File: docs/ingestion/compaction.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+
+### Dimension handling
+Apache Druid supports schema changes. Therefore, dimensions can be different across segments even if they are a part of the same data source. See [Different schemas among segments](../design/segments.md#different-schemas-among-segments). If the input segments have different dimensions, the resulting compacted segment includes all dimensions of the input segments. 
+
+Even when the input segments have the same set of dimensions, the dimension order or the data type of dimensions can be different. For example, the data type of some dimensions can be changed from `string` to primitive types, or the order of dimensions can be changed for better locality. In this case, the dimensions of recent segments precede that of old segments in terms of data types and the ordering because more recent segments are more likely to have the preferred order and data types.
```
Comment (2bethere) [inline docs/ingestion/compaction.md:64]:
```suggestion
Even when the input segments have the same set of dimensions, the dimension order or the data type of dimensions can be different. The dimensions of recent segments precede that of old segments in terms of data types and the ordering because more recent segments are more likely to have the preferred order and data types.
```
I think maybe deleting that sentence actually make it more clear...

File: docs/ingestion/compaction.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+|`context`|[Task context](./tasks.md#context)|No|
+|`granularitySpec`|Custom `granularitySpec` to describe the `segmentGranularity` and `queryGranularity` for the compacted segments. See [Compaction granularitySpec](#compaction-granularity-spec).|No|
+
+To control the number of result segments per time chunk, you can set [maxRowsPerSegment](../configuration/index.md#compaction-dynamic-configuration) or [numShards](../ingestion/native-batch.md#tuningconfig).
+
+> You can run multiple compaction tasks at the same time. For example, you can run 12 compaction tasks per month instead of running a single task for the entire year.
+
+A compaction task internally generates an `index` task spec for performing compaction work with some fixed parameters. For example, its `inputSource` is always the [DruidInputSource](native-batch.md#druid-input-source), and `dimensionsSpec` and `metricsSpec`include all dimensions and metrics of the input segments by default.
+
+Compaction tasks exit without doing anything and issue a failure status code:
```
Comment (2bethere) [inline docs/ingestion/compaction.md:108]:
```suggestion
Compaction tasks would exit without doing anything and issue a failure status code:
```

**Review by 2bethere — COMMENTED** [review]
Thanks for putting this up! I think it's a much-needed improvement. Added some comments.

**Review by sthetland — CHANGES_REQUESTED** [review]
Comments & suggestions below.. Looks good though! It makes compaction clearer.

File: docs/ingestion/compaction.md (hunk 3 above)
Comment (techdocsmith) [reply docs/ingestion/compaction.md:54]: @maytasm , I think by trying to clarify the behavior here, I made it more confusing. Can you help me with @2bethere 's questions?

File: docs/ingestion/data-management.md [hunk 7]
```diff
@@ -21,173 +21,9 @@ title: "Data management"
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
+Within the context of this topic Data management refers to Apache Druid's data maintenance capabilities for existing datasources. There are several options to help you keep your data relevant and to help your Druid cluster remain performant. For example updating, reingesting, adding lookups, reindexing, or deleting data.
```
Comment (suneet-s) [inline docs/ingestion/data-management.md:24]: nit: this line mentions "existing datasources" and the first topic in this page is `Adding new data`. I don't have a good idea yet on what the correct split should be yet.

File: docs/ingestion/compaction.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+As a general strategy, compaction is effective when you have data arriving out of chronological order resulting in lots of small segments. This often happens, for example, if you are appending data using `appendToExisting` for [native batch](./native_batch.md) ingestion. Conversely, if you are rewriting your data with each ingestion task, you don't need to use compaction.
+
+In some cases you can use compaction to reduce segment size. For example, if a misconfigured ingestion task creates oversized segments, you can create a compaction task to split the segment files into smaller, more optimally sized ones.
+
+See [Segment optimization](../operations/segment-optimization.md) for guidance to determine if compaction will help in your environment.
+
+
+## Types of segment compaction
```
Comment (suneet-s) [inline docs/ingestion/compaction.md:34]: why did you choose to call this `segment compaction` ? Don't have strong feelings right now, but just curious about the thought behind it

File: docs/ingestion/compaction.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+Query performance in Apache Druid depends on optimally sized segments. Compaction is one strategy you can use to optimize segment size for your Druid database. Compaction tasks read an existing set of segments for a given time interval and combine the data into a new "compacted" set of segments. The compacted segments are generally larger, but there are fewer of them. Here compaction increases performance because fewer segments require less the per-segment processing and the memory overhead for ingestion and for querying paths.
+
+As a general strategy, compaction is effective when you have data arriving out of chronological order resulting in lots of small segments. This often happens, for example, if you are appending data using `appendToExisting` for [native batch](./native_batch.md) ingestion. Conversely, if you are rewriting your data with each ingestion task, you don't need to use compaction.
```
Comment (suneet-s) [inline docs/ingestion/compaction.md:27]: This is slightly more nuanced. There are several reasons why you might find compaction useful, even when data does arrive in chronological order - like the parallelism in your parallel ingest task caused Druid to create many small segments. Also, the way this is worded makes me think only of streaming data (this might just be me), but this should also apply to batch ingestion when you "append" data

Perhaps a section that talks about all the reasons why someone would want to enabled compaction would be helpful.

File: docs/ingestion/compaction.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+In cases where you require more control over compaction, you can manually submit compaction tasks. For example:
+- Automatic compaction is too slow.
+- You want to force compaction for a specific time range.
+- You want to compact data out of chronological order.
+
+See [Setting up a manual compaction task](#setting-up-manual-compaction) for more about manual compaction tasks.
+
+
+## Data handling with compaction
+During compaction, Druid overwrites the original set of segments with the compacted set without modifying the data. During compaction Druid locks the segments for the time interval being compacted to ensure data consistency.
```
Comment (suneet-s) [inline docs/ingestion/compaction.md:48]: `without modifying the data.`

^ This used to be true till we added the ability to change queryGranularity with manual compaction. At the current time, this is the only way compaction will change the underlying data AFAIK

File: docs/ingestion/compaction.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+Apache Druid supports schema changes. Therefore, dimensions can be different across segments even if they are a part of the same data source. See [Different schemas among segments](../design/segments.md#different-schemas-among-segments). If the input segments have different dimensions, the resulting compacted segment includes all dimensions of the input segments. 
+
+Even when the input segments have the same set of dimensions, the dimension order or the data type of dimensions can be different. The dimensions of recent segments precede that of old segments in terms of data types and the ordering because more recent segments are more likely to have the preferred order and data types.
+
+If you want to use your own ordering and types, you can specify a custom `dimensionsSpec` in the compaction task spec.
+
+### Rollup
+Druid only rolls up the output segment when `rollup` is set for all input segments. See [Roll-up](../ingestion/index.md#rollup) for more details.
+You can check that your segments are rolled up or not by using [Segment Metadata Queries](../querying/segmentmetadataquery.md#analysistypes).
+
```
Comment (suneet-s) [inline docs/ingestion/compaction.md:74]: I think each one of these sections is currently describing the "what". Do you have a plan to talk about the "why"?

For example - You should change segmentGranularity in compaction if you ingested data and realized data for that time interval is sparse so having a larger segmentGranularity will result in better performance.

Similarly for queryGranularity - the why would be that you no longer need fine grained resolution in your data. There should be a big warning that tells users they can't go from coarser to finer granularity.

For dimension order - you'd want to change it to get better sorting and smaller segments.

I don't know what rollup does on it's own...

File: docs/ingestion/compaction.md (hunk 10 above)
Comment (suneet-s) [reply docs/ingestion/compaction.md:48]: Also, maybe we want to be more specific about what we mean by `modifying the underlying data`

What I mean by this is, compaction allows you to change the partitioning scheme of the underlying data, but for someone outside the database, they don't really notice this difference when querying the data.

However if queryGranularity is changed, they would notice the difference, because the timestamp column is changed. Sorry if this seems like a ramble, hopefully it helps explain what I'm thinking

**Comment by suneet-s** [comment]
I echo others comments on this PR. This is a huge improvement - thank you @techdocsmith ! I haven't verified the correctness of how exactly compaction works, or the details of the different tuning knobs

Some overall structural feedback (doesn't need to be addressed in this PR):
- I think the data management doc should be broken into a few separate docs. Seeing compaction pulled out of there - it feels like this would be a good landing page - that then points you to "getting data in", "Optimizing data", "Updating data"(maybe) and "Deleting data" This is obviously beyond the scope of this PR, but I think it's worth mentioning because it adds structure around how to think about data and managing data in Druid.
- Data management also talks about lookups, while the rest of the doc talks about datasources. This seemed a little out of place when I was reading locally. I don't have a suggestion for how to structure this right now, but wanted to surface it in case you had better ideas.
- The compaction page currently talks about the what. I wonder if it needs to be split into 2 pages (or sections), one that spells out the "why should I care/ I want to do..." a little bit more, and another that spells out "how do I do that". Maybe it can be intertwined in the same page?
- I really like the distinction between auto-compaction and manual compaction. However the page doesn't link to anything that tells me how to use auto-compaction, but it does link to something about manual compaction. Are there instructions for auto-compaction elsewhere?
- There are some known differences between auto-compaction and manual compaction. Support for queryGranularity is one right now. Do you think we should call this out in the section that talks about the differences between the 2. This is tricky, because it's like a gap in functionality - but it's a gotcha I think users will want to know about.

File: docs/ingestion/data-management.md (hunk 7 above)
Comment (techdocsmith) [reply docs/ingestion/data-management.md:24]: @suneet-s , that means adding new data to existing data sources. I will change that heading.

File: docs/ingestion/compaction.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+### Dimension handling
+Apache Druid supports schema changes. Therefore, dimensions can be different across segments even if they are a part of the same data source. See [Different schemas among segments](../design/segments.md#different-schemas-among-segments). If the input segments have different dimensions, the resulting compacted segment includes all dimensions of the input segments. 
+
+Even when the input segments have the same set of dimensions, the dimension order or the data type of dimensions can be different. The dimensions of recent segments precede that of old segments in terms of data types and the ordering because more recent segments are more likely to have the preferred order and data types.
+
+If you want to use your own ordering and types, you can specify a custom `dimensionsSpec` in the compaction task spec.
```
Comment (maytasm) [inline docs/ingestion/compaction.md:69]: I would say "If you want to force the ordering and types" or "If you want to ensure ordering and types is of a certain values" instead

File: docs/ingestion/compaction.md (hunk 11 above)
Comment (maytasm) [reply docs/ingestion/compaction.md:74]: You can't change rollup via compaction task right now.

To add to Suneet comment, for dimension order - you can change it to add or remove columns too. For example, for old data, you want no longer need x,y,z columns and wish to remove them or maybe you want to add a new aggregation on old data.

File: docs/ingestion/compaction.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+### Rollup
+Druid only rolls up the output segment when `rollup` is set for all input segments. See [Roll-up](../ingestion/index.md#rollup) for more details.
+You can check that your segments are rolled up or not by using [Segment Metadata Queries](../querying/segmentmetadataquery.md#analysistypes).
+
+## Setting up manual compaction
+
+To perform a manual compaction, you submit a compaction task. Compaction tasks merge all segments for the defined interval according to the following syntax:
+
+```json
```
Comment (maytasm) [inline docs/ingestion/compaction.md:79]: granularitySpec is missing from this example

File: docs/ingestion/compaction.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+|`tuningConfig`|[Parallel indexing task tuningConfig](./native-batch.md#tuningconfig)|No|
+|`context`|[Task context](./tasks.md#context)|No|
+|`granularitySpec`|Custom `granularitySpec` to describe the `segmentGranularity` and `queryGranularity` for the compacted segments. See [Compaction granularitySpec](#compaction-granularity-spec).|No|
+
+To control the number of result segments per time chunk, you can set [maxRowsPerSegment](../configuration/index.md#compaction-dynamic-configuration) or [numShards](../ingestion/native-batch.md#tuningconfig).
+
+> You can run multiple compaction tasks in parallel. For example, if you want to compact the data for a year, you are not limited to running a single task for the entire year. You can run 12 compaction tasks with month-long intervals.
+
+A compaction task internally generates an `index` task spec for performing compaction work with some fixed parameters. For example, its `inputSource` is always the [DruidInputSource](../native-batch.md#druid-input-source), and `dimensionsSpec` and `metricsSpec` include all dimensions and metrics of the input segments by default.
```
Comment (maytasm) [inline docs/ingestion/compaction.md:109]: I would change from "...dimensions and metrics of the input segments by default." to something like "...dimensions and metrics of the input segments when not given in the compaction spec".

File: docs/ingestion/compaction.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+      "queryGranularity": <time_period>
+    }
+    ...
+```
+
+`granularitySpec` takes the following keys:
+
+|Field|Description|Required|
+|-----|-----------|--------|
+|`segmentGranularity`|Time chunking period for the segment granularity. Defaults to 'null'. Accepts all [Query granularities](../querying/granularities.md).|No|
```
Comment (maytasm) [inline docs/ingestion/compaction.md:183]: Maybe explain the behavior of null here too

File: docs/ingestion/compaction.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+|`queryGranularity`|Time chunking period for the query granularity. Defaults to 'null'. Accepts all [Query granularities](../querying/granularities.md). Not supported for automatic compaction.|No|
```
Comment (maytasm) [inline docs/ingestion/compaction.md:184]: Maybe explain the behavior of null here too

File: docs/configuration/index.md [hunk 17]
```diff
@@ -842,24 +842,24 @@ A description of the compaction config is:
...
 |`maxRowsPerSegment`|Max number of rows per segment after compaction.|no|
-|`skipOffsetFromLatest`|The offset for searching segments to be compacted. Strongly recommended to set for realtime dataSources. |no (default = "P1D")|
+|`skipOffsetFromLatest`|The offset for searching segments to be compacted in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Strongly recommended to set for realtime dataSources. See [Data handling with compaction](../ingestion/compaction.md#data-handling-with-compaction)|no (default = "P1D")|
 |`tuningConfig`|Tuning config for compaction tasks. See below [Compaction Task TuningConfig](#compaction-tuningconfig).|no|
 |`taskContext`|[Task context](../ingestion/tasks.md#context) for compaction tasks.|no|
+|`granularitySpec`|Custom `granularitySpec` to describe the `segmentGranularity` for the compacted segments.|No|
```
[the remark this answers is not in the document]
Comment (maytasm) [reply docs/configuration/index.md:848]: Note that this is also different from manual compaction's granularitySpec described in #compaction-granularity-spec since the one here does not support `queryGranularity`

File: docs/ingestion/data-management.md [hunk 18]
```diff
@@ -21,173 +21,9 @@ title: "Data management"
...
 
-
-
-
-## Schema changes
-
-Schemas for datasources can change at any time and Apache Druid supports different schemas among segments.
-
-### Replacing segments
```
Comment (maytasm) [inline docs/ingestion/data-management.md:32]: Why is this section removed?

File: docs/ingestion/data-management.md [hunk 19]
```diff
@@ -21,173 +21,9 @@ title: "Data management"
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
+Within the context of this topic Data management refers to Apache Druid's data maintenance capabilities for existing datasources. There are several options to help you keep your data relevant and to help your Druid cluster remain performant. For example updating, reingesting, adding lookups, reindexing, or deleting data.
 
-
-
-
-## Schema changes
```
Comment (maytasm) [inline docs/ingestion/data-management.md:28]: Why is this section removed?

File: docs/ingestion/data-management.md [hunk 20]
```diff
@@ -21,173 +21,9 @@ title: "Data management"
...
-
-```
-foo_2015-01-01/2015-01-02_v1_0
-foo_2015-01-02/2015-01-03_v2_1
-foo_2015-01-03/2015-01-04_v1_2
-```
-
-In this case, queries may hit a mixture of `v1` and `v2` segments.
-
-### Different schemas among segments
```
Comment (maytasm) [inline docs/ingestion/data-management.md:80]: Why is this section removed?

File: docs/configuration/index.md [hunk 21]
```diff
@@ -842,24 +842,24 @@ A description of the compaction config is:
...
 
 An example of compaction config is:
 
 ```json
 {
-  "dataSource": "wikiticker"
+  "dataSource": "wikiticker",
```
Comment (maytasm) [inline docs/configuration/index.md:854]: Not have to be in this PR, but we should have a separate page dedicated to auto-compaction. This will allow us to go into more detail on why/how to use auto-compaction.

**Review by maytasm — CHANGES_REQUESTED** [review]

File: docs/ingestion/compaction.md [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+|`type`|Task type. Should be `compact`|Yes|
+|`id`|Task id|No|
+|`dataSource`|Data source name to compact|Yes|
+|`ioConfig`|I/O configuration for compaction task. See [Compaction I/O configuration](#compaction-io-configuration) for details.|Yes|
+|`dimensionsSpec`|Custom dimensions spec. The compaction task uses the specified dimensions spec if it exists instead of generating one.|No|
+|`metricsSpec`|Custom metrics spec. The compaction task uses the specified metrics spec rather than generating one.|No|
+|`segmentGranularity`|When set, the compaction task changes the segment granularity for the given interval.  Deprecated. Use `granularitySpec`. |No.|
+|`tuningConfig`|[Parallel indexing task tuningConfig](./native-batch.md#tuningconfig)|No|
+|`context`|[Task context](./tasks.md#context)|No|
+|`granularitySpec`|Custom `granularitySpec` to describe the `segmentGranularity` and `queryGranularity` for the compacted segments. See [Compaction granularitySpec](#compaction-granularity-spec).|No|
```
Comment (suneet-s) [inline docs/ingestion/compaction.md:103]: #10996 fails the compaction task if there is a conflict between the segmentGranularity specified in `granularitySpec` and `segmentGranularity`

I'm not sure what the best way to document that is in my PR, but since you are already refactoring this part of the docs, could you add something in this section surfacing that? Thanks!

File: docs/ingestion/compaction.md (hunk 3 above)
Comment (techdocsmith) [reply docs/ingestion/compaction.md:54]: Thanks @maytasm , I will add your example here.

File: docs/ingestion/data-management.md (hunk 18 above)
Comment (techdocsmith) [reply docs/ingestion/data-management.md:32]: it is 100% duplicated from ../design/segments.md

File: docs/ingestion/data-management.md (hunk 19 above)
Comment (techdocsmith) [reply docs/ingestion/data-management.md:28]: it is 100% duplicated from ../design/segments.md

File: docs/ingestion/data-management.md (hunk 20 above)
Comment (techdocsmith) [reply docs/ingestion/data-management.md:80]: it is 100% duplicated from ../design/segments.md

File: docs/configuration/index.md (hunk 21 above)
Comment (techdocsmith) [reply docs/configuration/index.md:854]: you are right, @maytasm . Filed a task to address this in another PR>

File: docs/ingestion/compaction.md (hunk 8 above)
Comment (techdocsmith) [reply docs/ingestion/compaction.md:34]: removed `segment`. I don't know that I had a specific reason in this case.

File: docs/ingestion/compaction.md (hunk 9 above)
Comment (techdocsmith) [reply docs/ingestion/compaction.md:27]: broke this paragraph into a section. made sure to mention parallel task, append to existing, and changing partitioning. This may need a little bit of revision.

File: docs/ingestion/compaction.md (hunk 11 above)
Comment (techdocsmith) [reply docs/ingestion/compaction.md:74]: I removed this rollup section since it seemed to not matter. I am adding the "why" into the Compaction Strategies section.

**Comment by suneet-s** [comment]
Docs failure looks legit

```
Could not find self anchor '#compaction-tuningconfig' in './build/ApacheDruid/docs/configuration/index.html'
Could not find './native_batch.md' linked from './build/ApacheDruid/docs/ingestion/compaction.html'
Could not find '../native-batch.md' linked from './build/ApacheDruid/docs/ingestion/compaction.html'
Could not find '../data-management.md' linked from './build/ApacheDruid/docs/ingestion/index.html'
Could not find '../compaction.md' linked from './build/ApacheDruid/docs/ingestion/index.html'
There are 5 issues
```

**Comment by techdocsmith** [comment]
Apologize for that @suneet-s , I fixed links and spelling in a later commit.

File: docs/ingestion/compaction.md [hunk 23]
```diff
@@ -22,36 +22,45 @@ description: "Defines compaction and automatic compaction (auto-compaction or au
...
+## Compaction strategies
+There are several cases to consider compaction for segment optimization:
+- With streaming ingestion, data can arrive out of chronological order creating lots of small segments.
+- If you append data using `appendToExisting` for [native batch](native-batch.md) ingestion creating suboptimal segments.
+- When you use `index_parallel` for parallel batch indexing and the parallel ingestion tasks create many small segments.
+- When a misconfigured ingestion task creates oversized segments.
 
-In some cases you can use compaction to reduce segment size. For example, if a misconfigured ingestion task creates oversized segments, you can create a compaction task to split the segment files into smaller, more optimally sized ones.
+By default, compaction does not modify the underlying data of the segments. However, there are cases when you may want to modify data during compaction to improve query performance:
+- If, after ingestion, you realize realize that data for the time interval is sparse, you can use compaction to increase the segment granularity.
```
Comment (maytasm) [inline docs/ingestion/compaction.md:35]: typo. realize twice

File: docs/ingestion/compaction.md [hunk 24]
```diff
@@ -22,36 +22,45 @@ description: "Defines compaction and automatic compaction (auto-compaction or au
...
+- Over time you don't need fine-grained granularity for older data so you use compaction to change older segments to a coarser query granularity. For example from `minute` to `hour`, or `hour` to `day`. You cannot go from coarser granularity to finer granularity.
```
Comment (maytasm) [inline docs/ingestion/compaction.md:36]: Should mentioned that the reason for this is for more rollup which result in less storage space

File: docs/ingestion/compaction.md [hunk 25]
```diff
@@ -62,16 +71,12 @@ Unless you modify the query granularity in the [granularity spec](#compaction-gr
...
-Apache Druid supports schema changes. Therefore, dimensions can be different across segments even if they are a part of the same data source. See [Different schemas among segments](../design/segments.md#different-schemas-among-segments). If the input segments have different dimensions, the resulting compacted segment includes all dimensions of the input segments. 
+Apache Druid supports schema changes. Therefore, dimensions can be different across segments even if they are a part of the same data source. See [Different schemas among segments](../design/segments.md#different-schemas-among-segments). If the input segments have different dimensions, the resulting compacted segment include all dimensions of the input segments. 
 
 Even when the input segments have the same set of dimensions, the dimension order or the data type of dimensions can be different. The dimensions of recent segments precede that of old segments in terms of data types and the ordering because more recent segments are more likely to have the preferred order and data types.
 
 If you want to use your own ordering and types, you can specify a custom `dimensionsSpec` in the compaction task spec.
 
-### Rollup
```
Comment (maytasm) [inline docs/ingestion/compaction.md]: Why is this removed?

Comment (maytasm) [reply docs/ingestion/compaction.md]: This is still true. What I meant to say is that compaction cannot change the `rollup` setting but the `rollup` setting on the original segments is important and determines how compaction works

File: docs/ingestion/compaction.md (hunk 13 above)
Comment (maytasm) [reply docs/ingestion/compaction.md:84]: Can you add granularitySpec to this json example

**Review by maytasm — CHANGES_REQUESTED** [review]

**Review by maytasm — APPROVED** [review]
